File: contexts/python/app/applog.py
"""The app's logger. Appends one JSON object per line to a log file — your
"log aggregation", such as it is. The log store reads this same file.
Override the path with the APP_LOG env var."""
import json
import os
import time


def log_path() -> str:
    return os.environ.get("APP_LOG") or os.path.join(
        os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "app.log"
    )


def emit(level: str, text: str) -> None:
    with open(log_path(), "a", encoding="utf-8") as f:
        f.write(json.dumps({"ts": time.time(), "level": level, "text": text}) + "\n")

File: contexts/typescript/app/applog.ts
// The app's logger. Appends one JSON object per line to a log file — your
// "log aggregation", such as it is. The log store reads this same file.
// Override the path with the APP_LOG env var.
import { appendFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

function logPath(): string {
  if (process.env.APP_LOG) return process.env.APP_LOG;
  const here = path.dirname(fileURLToPath(import.meta.url)); // .../app
  return path.join(here, "..", "app.log");
}

export function emit(level: string, text: string): void {
  appendFileSync(logPath(), JSON.stringify({ ts: Date.now() / 1000, level, text }) + "\n");
}
